### PR TITLE
Add a lightweight implementation for collecting Stream releated metrics

### PR DESCRIPTION
**Pull request by arunkumarchacko**

### CODE REVIEW

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java [hunk 1]
```diff
@@ -1856,7 +1865,7 @@ public Map<String, byte[]> getXAttrs(Path path) throws IOException {
   public Map<String, byte[]> getXAttrs(Path path, List<String> names) throws IOException {
     return GhfsStorageStatistics.trackDuration(
         storageStatistics,
-        GhfsStatistic.INVOCATION_XATTR_GET_MAP,
+        GhfsStatistic.INVOCATION_XATTR_GET_NAMED_MAP,
```
Comment (singhravidutt) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java:1868]: why this change?

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStream.java [hunk 2]
```diff
@@ -147,7 +141,7 @@ public void close() throws IOException {
               channel = null;
             }
           }
-
+          streamStats.close();
```
Comment (singhravidutt) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStream.java:144]: should this be part of finally block?

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java [hunk 3]
```diff
@@ -293,6 +287,8 @@ public synchronized void close() throws IOException {
             closeAPITrace(CLOSE_METHOD, stopwatch);
           }
 
+          streamStats.close();
```
Comment (singhravidutt) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java:290]: Don't we need a similar instrumentation for `close` call as well?

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java [hunk 4]
```diff
@@ -293,6 +287,8 @@ public synchronized void close() throws IOException {
...
+          seekStreamStats.close();
```
Comment (singhravidutt) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java:291]: what about closing the statistics stream in finally block?

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java [hunk 5]
```diff
@@ -337,12 +333,12 @@ private void readAPITrace(
...
   }
 
-  private void seekAPITrace(String method, Stopwatch stopwatch, long pos) {
+  private void seekAPITrace(String method, long startTimeNs, long pos) {
     if (isTraceLoggingEnabled) {
       Map<String, Object> jsonMap = new HashMap<>();
       addLogProperty(METHOD, method, jsonMap);
       addLogProperty(GCS_PATH, gcsPath, jsonMap);
-      addLogProperty(DURATION_NS, stopwatch.elapsed(TimeUnit.NANOSECONDS), jsonMap);
+      addLogProperty(DURATION_NS, getElapsedMillisFromStartTime(startTimeNs), jsonMap);
```
Comment (singhravidutt) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java:341]: DURATION_MS?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamIntegrationTest.java [hunk 6]
```diff
@@ -267,21 +267,17 @@ public void operation_durationMetric_tests() throws Exception {
...
+    TestUtils.verifyDurationMetric(stats, STREAM_READ_OPERATIONS, 0);
 
     in.seek(0);
 
-    TestUtils.verifyDurationMetric(stats, STREAM_READ_SEEK_OPERATIONS, 3);
-
-    TestUtils.verifyDurationMetric(stats, STREAM_READ_CLOSE_OPERATIONS, 1);
+    TestUtils.verifyDurationMetric(stats, STREAM_READ_SEEK_OPERATIONS, 0);
 
-    TestUtils.verifyDurationMetric(stats, STREAM_READ_CLOSE_OPERATIONS, 2);
```
Comment (singhravidutt) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamIntegrationTest.java:282]: why we have removed close operation  verification?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamIntegrationTest.java [hunk 7]
```diff
@@ -267,21 +267,17 @@ public void operation_durationMetric_tests() throws Exception {
     assertThat(in.read(1, value, 1, 1)).isEqualTo(1);
     assertThat(value).isEqualTo(expected);
 
-    // One seek is to seek to the position and another seek to seek back to the original position.
-    TestUtils.verifyCounter(stats, STREAM_READ_SEEK_OPERATIONS, 2);
+    TestUtils.verifyCounter(stats, STREAM_READ_SEEK_OPERATIONS, 0);
     TestUtils.verifyCounter(stats, STREAM_READ_SEEK_BACKWARD_OPERATIONS, 2);
 
-    TestUtils.verifyDurationMetric(stats, STREAM_READ_OPERATIONS, 2);
+    TestUtils.verifyDurationMetric(stats, STREAM_READ_OPERATIONS, 0);
```
Comment (singhravidutt) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamIntegrationTest.java:273]: instead of verifying individual metric to be `0` what about verifying that nothing will be captured unless stream is closed?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopSyncableOutputStreamIntegrationTest.java [hunk 8]
```diff
@@ -88,7 +88,7 @@ public void hsync() throws Exception {
     assertThat(gcsFsIHelper.readFile(path)).isEqualTo(expected);
 
     TestUtils.verifyDurationMetric(
-        stats, GhfsStatistic.STREAM_WRITE_CLOSE_OPERATIONS, expected.length);
+        stats, GhfsStatistic.STREAM_WRITE_CLOSE_OPERATIONS, expected.length + 1);
```
Comment (singhravidutt) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopSyncableOutputStreamIntegrationTest.java:91]: why?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamIntegrationTest.java [hunk 9]
```diff
@@ -130,7 +130,7 @@ public void time_statistics() throws Exception {
     fout.close();
 
     TestUtils.verifyDurationMetric(stats, GhfsStatistic.STREAM_WRITE_CLOSE_OPERATIONS, 1);
-    TestUtils.verifyDurationMetric(stats, GhfsStatistic.STREAM_WRITE_OPERATIONS, 4);
+    TestUtils.verifyDurationMetric(stats, GhfsStatistic.STREAM_WRITE_OPERATIONS, 2);
```
Comment (singhravidutt) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamIntegrationTest.java:133]: why it reduced?

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GhfsStreamStats.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        toMillis(minLatencyNs),
+        toMillis(maxLatencyNs),
+        toMillis(totalNs),
+        operationCount,
+        context);
+    this.totalNs = 0;
+    this.operationCount = 0;
+  }
+
+  void updateWriteStreamStats(int len, long start) {
```
[the remark this answers is not in the document]
Comment (arunkumarchacko) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GhfsStreamStats.java:55]: yes, we could have. Both would have worked. I think this one might be a bit better since we are explicit about the method name.

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java (hunk 3 above)
Comment (arunkumarchacko) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java]: close() wont be called this frequently. So, a more readable implementation is better.

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java (hunk 1 above)
Comment (arunkumarchacko) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java:1868]: The earlier implementation was wrong.  it should have been NAMED_MAP.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamIntegrationTest.java (hunk 7 above)
Comment (arunkumarchacko) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamIntegrationTest.java:273]: some metrics will still be captured before close. Like backward/forward seek.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamIntegrationTest.java (hunk 9 above)
Comment (arunkumarchacko) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStreamIntegrationTest.java:133]: there was a bug in the verification code. (Plz refer TestUtils.java). I fixed that. The correct value should have been 2. One for each write() in the test.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopSyncableOutputStreamIntegrationTest.java (hunk 8 above)
Comment (arunkumarchacko) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopSyncableOutputStreamIntegrationTest.java:91]: One write for each hsync() and one for the actual file. This was masked due to the bug in the test verification code.

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java (hunk 4 above)
Comment (arunkumarchacko) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java]: close() do not throw exception in normal case. Agree, it would be safer. Changed.

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStream.java (hunk 2 above)
Comment (arunkumarchacko) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopOutputStream.java]: Done

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java (hunk 5 above)
Comment (arunkumarchacko) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java:344]: Done. Good catch.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamIntegrationTest.java (hunk 6 above)
Comment (arunkumarchacko) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStreamIntegrationTest.java:282]: Same as the other instance. There was a bug in the test verification. There is only 1 close operation and it is checked below.

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GhfsStreamStats.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (singhravidutt) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GhfsStreamStats.java:55]: wouldn't it cause maintenance issue?

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java (hunk 3 above)
Comment (singhravidutt) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java]: IMO, It would be better to follow one strategy all across.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopSyncableOutputStreamIntegrationTest.java (hunk 8 above)
Comment (singhravidutt) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopSyncableOutputStreamIntegrationTest.java:91]: mind adding a comment here?

**Review by singhravidutt — APPROVED** [review]

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java (hunk 3 above)
Comment (arunkumarchacko) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java]: The other pattern is used at 20+ places. It is useful to have code reuse.
